_posts/2023-02-28-kani-internship-projects-2022-stubbing.md

```diff
@@ -0,0 +1,361 @@
+
+#[cfg(kani)]
+#[kani::proof]
+#[kani::stub(rand::random, stub_random)]
+fn encrypt_then_decrypt_is_identity() {
+    let data: u32 = kani::any();
+    let encryption_key: u32 = rand::random();
+    let encrypted_data = data ^ encryption_key;
+    let decrypted_data = encrypted_data ^ encryption_key;
+    assert_eq!(data, decrypted_data);
+}
+```
+
+Here, the `#[kani::stub(rand::random, stub_random)]` attribute indicates to Kani that it should replace `rand::random` with the stub `stub_random`.
+Note that this is a sound assumption to do: `rand::random` is expected to return any `u32` value, just like `kani::any`.
+
+Now, let's run it through Kani:
+
+```bash
+cargo kani --enable-unstable --enable-stubbing --harness encrypt_then_decrypt_is_identity
+```
+
+The verification result is composed of a single check: the assertion corresponding to `assert_eq!(data, decrypted_data)`.
+
+```
+RESULTS:
+Check 1: encrypt_then_decrypt_is_identity.assertion.1
+         - Status: SUCCESS
+         - Description: "assertion failed: data == decrypted_data"
+         - Location: src/main.rs:18:5 in function encrypt_then_decrypt_is_identity
+
+
+SUMMARY:
+ ** 0 of 1 failed
+
+VERIFICATION:- SUCCESSFUL
+```
+
+Kani verifies the assertion successfully, avoiding any issues that appear if we attempt to verify the code without stubbing.
+
+### Another example: stubbing `factorial`
+
+Let's see another example where we compute the [factorial](https://en.wikipedia.org/wiki/Factorial) of a number `n`.
+Note that the `factorial` function uses the method `checked_mul` to prevent overflow errors.
+We're interested in verifying the correctness of the `factorial` calculation.
+
+```rust
+fn factorial(n: u64) -> Option<u64> {
+    if n == 0 {
+        return Some(1);
+    }
+    n.checked_mul(factorial(n - 1).unwrap())
+}
+
+#[cfg(kani)]
+#[kani::proof]
+fn verify_factorial() {
+    let n = kani::any();
+    kani::assume(n > 0);
+    if let Some(value) = factorial(n) {
+        assert_eq!(value, n * factorial(n - 1).unwrap());
+    }
+}
+```
+
+If we run this example through Kani, we'll see the following output:
+
+```
+Unwinding recursion factorial iteration 1
+Unwinding recursion factorial iteration 2
+Unwinding recursion factorial iteration 3
+Unwinding recursion factorial iteration 4
+Unwinding recursion factorial iteration 5
+Unwinding recursion factorial iteration 6
+Unwinding recursion factorial iteration 7
+Unwinding recursion factorial iteration 8
+Unwinding recursion factorial iteration 9
+Unwinding recursion factorial iteration 10
+# -- snip --
+```
+
+In this case, it doesn't look like Kani will terminate anytime soon due to the recursive definition of the `factorial` function.
+The code under verification shows bad performance because the underlying engine attempts to unwind all possible recursive calls.
+
+But we can avoid the recursion by pre-computing the values that fit into a `u64` value and replacing `factorial` with another function that gets those values.
```

Comment:
Perhaps add a footnote that says that we only need 21 values because beyond 21, the factorial computation overflows.